### PR TITLE
*: Merge global temporary table into release-5.1

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

This branch is maintained for a POC of a customer, it provides the global temporary table feature based on the 5.1 branch.
And it meant to keep update with the 5.1 branch, plus the global temporary table feature.

Problem Summary:

### What is changed and how it works?

What's Changed:

Initially, this branch is generated by cherry pick those PRs based on the 5.1 branch:

```
commit 811253785c0ed3ac84904517a73a2667f70f19f3
Author: 王超 <cclcwangchao@hotmail.com>
Date:   Mon Jun 7 21:06:28 2021 +0800

    planner, executor: add stale read compatibility for temporary table (#25206)


commit d5a0d176db39bae07f4584df119e81170dade1eb
Author: 王超 <cclcwangchao@hotmail.com>
Date:   Wed Jun 9 13:18:29 2021 +0800

    executor: add history read compatibility for temporary table (#25237)

commit cbb6f4e64cb9e46f3db3d36c8c1abbd80ebd5b73
Author: tiancaiamao <tiancaiamao@gmail.com>
Date:   Wed Jun 9 18:48:30 2021 +0800

    *: add global system variable `tmp_table_size` (#24827)

commit 0ed8eafa9bf269eb2c3974ce67184aa05fbabc38
Author: 王超 <cclcwangchao@hotmail.com>
Date:   Wed Jun 9 21:14:29 2021 +0800

    executor: add tablesample compatibility for temporary table (#25288)

commit 5a1606740c5225175b5b5a64bad54cc83d3ff18c
Author: Howie <haoweili35@gmail.com>
Date:   Fri Jun 11 16:50:35 2021 +0800

    ddl: add create table like check for temporary table (#25082)

commit bd11917ca096374ca1d7915d5985f49cb111be3e
Author: tiancaiamao <tiancaiamao@gmail.com>
Date:   Tue Jun 15 14:08:37 2021 +0800

    bindinfo,planner: report error when creating sql binding on temporary table (#25058)

commit ed686d1963f472b813c528543fee729a629242fb
Author: tiancaiamao <tiancaiamao@gmail.com>
Date:   Thu Jun 17 16:20:37 2021 +0800

    executor: fix a stupid mistake in `filterTemporaryTableKeys()` function (#25521)

commit 9f18723e6af2e1bb10fb013a75c2397a7a575e2c
Author: tiancaiamao <tiancaiamao@gmail.com>
Date:   Mon Jun 21 16:46:53 2021 +0800

    *: fix bug that write on temporary table send request to TiKV (#25535)
```

How it Works:

This draft issue can make us easier to know the different between this branch and 5.1.
The branch will be maintained like this:

* Cherry-pick temporary table related bug fixes from the master branch
* Merge 5.1 branch regularly

Until the POC is over and we finally provide a official release branch that contains the global temporary table feature.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note

### CODE REVIEW

**Review by djshow832 — APPROVED** [review]

File: errno/errname.go [hunk 1]
```diff
@@ -925,6 +925,7 @@ var MySQLErrName = map[uint16]*mysql.ErrMessage{
 	ErrForUpdateCantRetry:            mysql.Message("[%d] can not retry select for update statement", nil),
 	ErrAdminCheckTable:               mysql.Message("TiDB admin check table failed.", nil),
 	ErrOptOnTemporaryTable:           mysql.Message("`%s` is unsupported on temporary tables.", nil),
+	ErrDropTableOnTemporaryTable:     mysql.Message("`drop global temporary table` can only drop global temporary table", nil),
```
Comment (bb7133) [inline errno/errname.go:928]:
```suggestion
	ErrDropTableOnTemporaryTable:     mysql.Message("`drop global temporary table` applies on global temporary table only.", nil),
```

File: errors.toml [hunk 2]
```diff
@@ -1156,6 +1161,11 @@ error = '''
 `%s` is unsupported on temporary tables.
 '''
 
+["planner:8007"]
+error = '''
+`drop global temporary table` can only drop global temporary table
```
Comment (bb7133) [inline errors.toml:1166]: Ditto

Comment (tiancaiamao) [reply errors.toml]: Done

**Review by bb7133 — APPROVED** [review]

**Pull request merged**